### PR TITLE
Refactor complex comparison functions

### PR DESCRIPTION
**Pull request by ndgrigorian**
This PR refactors complex comparison functions into a new header file, `utils\math_utils.hpp`.

- [X] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [ ] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

File: dpctl/tensor/libtensor/include/utils/math_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+std::complex<T> max_complex(const std::complex<T> &in1,
+                            const std::complex<T> &in2)
+{
+    T real1 = std::real(in1);
+    T real2 = std::real(in2);
+    T imag1 = std::imag(in1);
+
+    bool is_greater =
+        (real1 == real2) ? (imag1 > std::imag(x2)) : real1 > real2;
+    return (std::isnan(real1) || std::isnan(imag1) || greater_c) ? in1 : in2;
```
Comment (vtavana) [inline dpctl/tensor/libtensor/include/utils/math_utils.hpp:87]: Following your suggestion in #1316 , I noticed it does not return `in2` if imaginary part of `in2` is `NaN`.
Something similar to the following works fine.

```suggestion
            T imag2 = std::imag(in2);
            
            if (std::isnan(real1) || std::isnan(imag1))
                return in1;
            else if (std::isnan(real2) || std::isnan(imag2))
                return in2;
            else if (real1 == real2) 
                return imag1 > imag2 ? in1 : in2;
            else
                return real1 > real2 ? in1 : in2;
```

**Comment by oleksandr-pavlyk** [comment]
For some reason changes a causing a new failure in array API test suite: 

> array_api_tests/test_creation_functions.py ........F.....                [  3%]

This failure is not present in the main branch.

**Comment by ndgrigorian** [comment]
> For some reason changes a causing a new failure in array API test suite:
> 
> > array_api_tests/test_creation_functions.py ........F.....                [  3%]
> 
> This failure is not present in the main branch.

I saw this too. It's in linspace. It doesn't seem to be related to this PR, but worth looking at.

**Comment by oleksandr-pavlyk** [comment]
Yes, the "test_creation_functions.py" failure is sporadic, I just reran the workflow steps and this particular test passed.

**Review by oleksandr-pavlyk — APPROVED** [review]

**Comment by oleksandr-pavlyk** [comment]
I just reproduced the failure locally:

```
E           hypothesis.errors.Flaky: Hypothesis test_linspace(num=3866, dtype=dtype('float32'), endpoint=False, data=data(...)) produces unreliable results: Falsified on the first call but did not on a subsequent one
E           Falsifying example: test_linspace(
E               num=3866,
E               dtype=dtype('float32'),
E               endpoint=False,
E               data=data(...),
E           )
E           Draw 1 (start): 9.999999747378752e-06
E           Draw 2 (stop): -2.220446049250313e-16
E           Draw 3 (kw): {'dtype': dtype('float32'), 'endpoint': False}
E           Failed to reproduce exception. Expected:
E           >   ???
```

with the underlying cause being the following failure in "_copy_from_numpy_into(dst, np_ary)":

```
E           >       ti._copy_numpy_ndarray_into_usm_ndarray(
E                       src=src_ary, dst=dst, sycl_queue=copy_q
E                   )
E           E       RuntimeError: Native API failed. Native API returns: -30 (PI_ERROR_INVALID_VALUE) -30 (PI_ERROR_INVALID_VALUE)
```

**Pull request merged**